### PR TITLE
Strategy to group unbound view connections 

### PR DESCRIPTION
**Pull request by mmandlis**
(of particles added by search-token-to-particle strategy)

### CODE REVIEW

File: runtime/strategies/group-view-connections.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+// Code distributed by Google as part of this project is also
+// subject to an additional IP rights grant found at
+// http://polymer.github.io/PATENTS.txt
+
+let assert = require('chai').assert;
+let {Strategy} = require('../../strategizer/strategizer.js');
+let Recipe = require('../recipe/recipe.js');
+let RecipeWalker = require('../recipe/walker.js');
+
+module.exports = class GroupViewConnections extends Strategy {
```
Comment (dstockwell) [inline runtime/strategies/group-view-connections.js]: I think we should start adding documentation about what individual strategies do in general. I'm a bit lost coming back to this one.

File: runtime/strategies/group-view-connections.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2017 Google Inc. All rights reserved.
+// This code may only be used under the BSD style license found at
+// http://polymer.github.io/LICENSE.txt
+// Code distributed by Google as part of this project is also
+// subject to an additional IP rights grant found at
+// http://polymer.github.io/PATENTS.txt
+
+let assert = require('chai').assert;
```
Comment (dstockwell) [inline runtime/strategies/group-view-connections.js]:
```
... require('assert')
```

File: runtime/strategies/group-view-connections.js (hunk 1 above)
Comment (mmandlis) [reply runtime/strategies/group-view-connections.js]: totally agree, we should have documentation
this one is for particles that were added based on search tokens, and don't have any bindings between their viewconnections. 
so i'm trying to connect in viewconnections with out view connections from different particles as much as possible.
this is very speculative, but i figured we should start somewhere...

**Comment by mmandlis** [comment]
I made some updates, can you, please, take another look?
I added a CombinedStrategy that runs a list of strategies together, and use it to run
SearchTokensToParticles and GroupViewConnections to avoid explosion of options.
Or we can discuss our meeting tomorrow, if needed.

File: runtime/strategies/search-tokens-to-particles.js [hunk 3]
```diff
@@ -23,23 +24,17 @@ module.exports = class SearchTokensToParticles extends Strategy {
...
-    var results = Recipe.over([...generated, ...terminal], new class extends RecipeWalker {
+    class Walker extends RecipeWalker {
       onRecipe(recipe) {
         if (!recipe.search || !recipe.search.unresolvedTokens.length) {
           return;
         }
 
         let specsByToken = {};
         for (let token of recipe.search.unresolvedTokens) {
+          //for (let spec of this.particlesFinder(token)) {
```
Comment (dstockwell) [inline runtime/strategies/search-tokens-to-particles.js]: remove?

File: runtime/strategies/combined-strategy.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    assert(strategies.length > 1, 'Strategies must contain at least 2 elements.');
+    this._strategies = strategies;
+    this._strategies.forEach(strategy => assert(strategy.walker));
+    assert(this._strategies[0].getResults);
+  }
+  async generate(strategizer) {
+    let results = this._strategies[0].getResults(strategizer);
+    let totalResults = [];
+    this._strategies.forEach(strategy => {
+      totalResults.forEach(r => r.final = true);
```
Comment (dstockwell) [inline runtime/strategies/combined-strategy.js]: I don't quite understand what's going on with `getResults` and `.final`, can you explain the problem? I'm hoping we can find a way to build `CombinedStrategy` without changes to strategizer.

File: runtime/strategies/search-tokens-to-particles.js (hunk 3 above)
Comment (mmandlis) [reply runtime/strategies/search-tokens-to-particles.js]: done

File: runtime/strategies/combined-strategy.js (hunk 4 above)
Comment (mmandlis) [reply runtime/strategies/combined-strategy.js]: I got rid of .final (strategy-explorer populates missing derivation parents instead)
getResults abstraction is needed for the combined strategy to pass each strategy whatever recipes the it needs to operate on - for most of them is it the strategizer.generated, but search-tokens-to-particles needs strategizer.terminal too

**Comment by mmandlis** [comment]
PTAL

File: runtime/strategies/combined-strategy.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+module.exports = class CombinedStrategy extends Strategy {
+  constructor(strategies) {
+    super();
+    assert(strategies.length > 1, 'Strategies must contain at least 2 elements.');
+    this._strategies = strategies;
+    this._strategies.forEach(strategy => assert(strategy.walker));
+    assert(this._strategies[0].getResults);
+  }
+  getLeafs(results) {
```
Comment (dstockwell) [inline runtime/strategies/combined-strategy.js]: would that be `getLeaves`?

**Pull request merged**